walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,64 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.Collections;
+import com.google.sps.Event;
+import com.google.sps.MeetingRequest;
+import com.google.sps.TimeRange;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Do not provide any time options if the meeting requested is longer than a whole day. 
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    Collection<TimeRange> times = new ArrayList<>();
+
+    // Start and end markers of a time range that is available for a meeting. 
+    int start = TimeRange.START_OF_DAY;
+    int end = TimeRange.START_OF_DAY;
+
+    for (Event event : events) {
+
+      // Do not use event to determine time range options if none of the event attendees will be
+      // attending the requested meeting. 
+      if (Collections.disjoint(event.getAttendees(), request.getAttendees())) {
+        continue;
+      }
+
+      TimeRange eventTime = event.getWhen();
+
+      // "Skips" over events that start at the same time and places start marker at end of event.
+      if (start == eventTime.start()) {
+        start = eventTime.end();
+        continue;
+      }
+
+      end = eventTime.start();
+      
+      TimeRange time = TimeRange.fromStartEnd(start, end, false);
+
+      // Only add time range if it's long enough to host the meeting.
+      if (time.duration() >= request.getDuration()) {
+        times.add(time);
+      }
+
+      // Moves marker to set the start of the next available time range. 
+      if (start < eventTime.end()) {
+        start = eventTime.end();
+        end = eventTime.end();
```

Comment:
I don't quite understand the purpose of this line `end=eventTime.end()`.Could you please explain it a bit?

Reply:
I put it in so that end wouldn't be earlier in the day than start (it just made more sense to me logically), but looking through the code I don't actually think it's necessary. I'll take it out, thanks!